sql/parser/builtins.go

```diff
@@ -963,15 +980,20 @@ func datumToString(datum Datum) (string, error) {
 
 var replaceSubRe = regexp.MustCompile(`\\[&1-9]`)
 
-func regexpReplace(s, pattern, to, sqlFlags string) (Datum, error) {
+func regexpReplace(ctx EvalContext, s, pattern, to, sqlFlags string) (Datum, error) {
 	pattern, global, err := regexpEvalFlags(pattern, sqlFlags)
 	if err != nil {
 		return nil, err
 	}
 
-	patternRe, err := regexp.Compile(pattern)
-	if err != nil {
-		return nil, err
+	var patternRe *regexp.Regexp
+	if patternRe = ctx.ReCache.Lookup(pattern); patternRe == nil {
```

Comment:
I'd move this logic into the cache directly; I think you want to hold the mutex during the full lookup-and-update cycle. Also, is it worth caching regexps that produce compilation errors?